### DISCUSSION TITLE
Run a code when an asset failed

### DISCUSSION THREAD

**Discussion by dridk** [dagster-io/dagster#20381] [Q&A] [answered]
hi, 

can I run some code when an asset failed ? 

I have an asset which create a temporary directory. This directory is removed at the end of the asset materialization. but If asset materialization failed before the instructions it would not.

**Comment by cody-scott** [comment]
Could you not catch the error, do clean up then raise it again so it surfaces into the ui?

**Comment by cmpadden** [comment, marked as answer]
Hi @dridk - this can be accomplished in a few ways.

It is in fact possible to use a sensor to run a job on the failure of an asset materialization -- see the run status sensors documentation for more information.

https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#run-status-sensors

However, to @cody-scott's point, you may be better off either catching the exception and handling the cleanup in your asset, or I think my preferred approach which is to use the `tempfile` module so that the temporary directory is automatically deleted once outside of the scope of the `with` block.

```
import tempfile

@asset
def example_asset():
    with tempfile.TemporaryDirectory() as tmp:
        # perform operations reliant on temporary directory
```

**Reply by dridk** [reply]
Hi, thx for your quick reply. 
Unfortunately exception and tempdir lib doesn't fit with my problem. 
I am using a graph asset with different @op writing on this temp directory. 

Can I catch exception from different op within the asset_graph?

```python

@op 
def download_docs(): 
    pwd =Mkdir("tempdir") # create temp dir
    Download(many docs) 
    Return pwd

@op
def convert_to_htmls(pwd):
    .....
    return pwd

@op 
def convert_to_texts(pwd):
    ....
    return pwd

@op
def clean(pwd): 
     rmdir() # drop temp dir
     ....
     Return dataframe

@graph_asset()
def asset():
    pwd=mkdir()
    pwd = download_docs(pwd)
    pwd = convert_htmls(pwd)
    ....

```

**Reply by cmpadden** [reply]
I see - what you're describing seems similar to this discussion: https://github.com/dagster-io/dagster/discussions/16166#discussioncomment-6855132

Does that seem like a viable solution?

**Reply by dridk** [reply]
Yea, it is !  Just discovered the is_required Args. 

Thanks
